### ISSUE TITLE
RyuJIT and SIMD - suboptimal codegen when performing bitmask comparisons

### ISSUE THREAD

**Issue by GrabYourPitchforks** [dotnet/coreclr#18603]
```cs
    private static bool TestMask(ref Vector<byte> a, ref Vector<byte> b)
    {
        return (a & b) == Vector<byte>.Zero;
    }
```

This generates the codegen (x64):

```asm
00007FF82F4D1524  vmovupd     ymm0,ymmword ptr [rsi]  
00007FF82F4D1529  vmovupd     ymm1,ymmword ptr [rdi]  
00007FF82F4D152E  vpand       ymm0,ymm0,ymm1  
00007FF82F4D1533  vptest      ymm0,ymm0  
00007FF82F4D1538  sete        al 
```

The `vpand` instruction is unnecessary, as the comparison line could read simply `vptest ymm0,ymm1`, which sets both _ZF_ and _CF_ (see http://www.felixcloutier.com/x86/PTEST.html).

**Comment by GrabYourPitchforks** [comment]
I guess technically we could get rid of the second `vmovupd` as well since the comparison instruction could read `vptest ymm0,ymmword ptr [rdi]` (second operand using rmreg memory syntax).

**Comment by AndyAyersMS** [comment]
Probably could do some tweaking to `Lowering::LowerSIMD` here. Looks like SETCC removal is inhibited by the cast:
```
N009 ( 19, 16) [000011] ---XG-------              *  CAST      int <- bool <- int $1c0
N007 (  2,  2) [000007] ------------              |  /--*  SIMD      simd32 int init $181
N006 (  1,  1) [000006] -c----------              |  |  \--*  CNS_INT   int    0 $40
N008 ( 18, 14) [000008] ---XG-------              \--*  SIMD      int    ubyte == $182
N004 (  7,  5) [000004] ---XG-------                 |  /--*  IND       simd32
N003 (  1,  1) [000003] ------------                 |  |  \--*  LCL_VAR   byref  V01 arg1         u:2 (last use) $81
N005 ( 15, 11) [000005] ---XG-------                 \--*  SIMD      simd32 ubyte & $180
N002 (  7,  5) [000002] ---XG-------                    \--*  IND       simd32
N001 (  1,  1) [000001] ------------                       \--*  LCL_VAR   byref  V00 arg0         u:2 (last use) $80
```
This might make a good up for grabs issue....

**Comment by jakobbotsch** [comment]
I can try to take a look at this. To be clear, `Lowering::LowerSIMD` is for the containment of `ymmword ptr [rdi]`, and SETCC removal would be somewhere else, right?

**Comment by mikedn** [comment]
It looks to me that there are multiple issues being discussed here:
* lack of AND+CMP combining for SIMD. This is something that the JIT does for integral types (see `Lowering::OptimizeConstCompare` in lower.cpp) but not for SIMD types. In theory this should be a simple optimization but I suspect that things are going to be a bit more complicated in the case of SIMD  at least because `ptest` is not available before SSE 4.1 and there's no actual SIMD intrinsic for it, codegen simply recognizes a zero equality check. Contrast this with `OptimizeConstCompare` where `GT_TEST_EQ/NE` is being used.
* lack of memory operand containment. AFAIR this is a general issue of SIMD intrinsics but it's likely to be more complicated in this case, in part due to the fact that `ptest` generation is deferred until codegen as already mentioned above,
* an useless cast being generated after SIMD equality. This is likely a side effect of SIMD intrinsics being imported from method calls that need some special handling in the case of small return types. It's obviously not necessary because SIMD equality always produces 0/1. It would be nice to get rid of it but I'm not sure what @AndyAyersMS refers to when talking about SETCC removal.

**Comment by jakobbotsch** [comment]
With respect to 1, what would be the preferred way to solve it? I guess we could match the tree in lowering, eg. transform a tree such as

```
        SIMD EQ
        /     \
 SIMD AND      SIMD 0 constant
 /      \
A        B
```
into
```
    SIMD AND
    /      \
   A        B
```

However, I don't find it particularly pretty having to rely on code-gen generating `ptest` and keeping the conditions (eg. SSE 4.1) up to date in two places.

**Comment by jakobbotsch** [comment]
Is it feasible to add a new intrinsic ID for `ptest`? That way we could lower
```
        SIMD EQ
        /     \
 SIMD AND      SIMD 0 constant
 /      \
A        B
```

to something like
```
 SIMD  PTEST
 /      \
A        B
```
for SSE 4.1+. That would make the containment problem easier to fix as well.

**Comment by mikedn** [comment]
> However, I don't find it particularly pretty having to rely on code-gen generating ptest and keeping the conditions (eg. SSE 4.1) up to date in two places.

Yes, it's best if lowering produces IR that resembles the final code instead of relying on codegen to perform additional transforms/optimizations. It's not always possible/feasible/beneficial but in this case it probably can be done easily by simply adding a SIMD intrinsic that represents `ptest` and have lowering generate `SIMDIntrinsicTest` followed by `SETCC`/`JCC` when possible.

cc @CarolEidt

**Comment by tannergooding** [comment]
@mikedn, @CarolEidt. Whatever we do, it might be beneficial to do it in a way that can also be consumed via Hardware Intrinsics, since similar logic is/will be required there.

**Comment by mikedn** [comment]
> Whatever we do, it might be beneficial to do it in a way that can also be consumed via Hardware Intrinsics, since similar logic is/will be required there.

I'm not sure how that's relevant, with HW intrinsics the developer has direct access to `ptest`.

**Comment by tannergooding** [comment]
> the developer has direct access to ptest.

Yes, via several different APIs that return a `bool`. The logic for folding the result of the `ptest` and the subsequent `cmov`, `setcc`, `jcc`, etc should be shared if possible.

**Comment by mikedn** [comment]
> The logic for folding the result of the ptest and the subsequent cmov, setcc, jcc, etc should be shared if possible.

That's a different thing. And there's a separate issue for it: #17073

**Comment by CarolEidt** [comment]
I believe that the intrinsics for ptest are already defined here https://github.com/dotnet/coreclr/blob/master/src/System.Private.CoreLib/shared/System/Runtime/Intrinsics/X86/Sse41.cs#L569 and below, but not yet implemented, so it would be a matter of implementing the SSE4.1 intrinsics, and then adding the support in `Lowering` to do the transformation. (And, separately, adddress #17073)

**Comment by mikedn** [comment]
> I believe that the intrinsics for ptest are already defined here 

Those are HW intrinsics, this is about SIMD intrinsics. Though I wonder if it would be possible to just reuse the HW intrinsic in this case, hmm...

**Comment by tannergooding** [comment]
@CarolEidt, SSE4.1 should be fully implemented

**Comment by CarolEidt** [comment]
@tannergooding - thanks; for some reason I missed seeing the implementation for them when I looked.
@mikedn - Yes, I would thank that, by the time we're in `Lowering` we should just be able to check for availability and then use the HW intrinsics.

**Comment by mikedn** [comment]
> Yes, I would thank that, by the time we're in Lowering we should just be able to check for availability and then use the HW intrinsics.

Sounds good but then we need to deal with #17073 first because existing `LowerSIMD` relies on `SETCC`/`JCC` and HW intrinsics don't support that.

**Comment by mikedn** [comment]
> if only to get some guidance whether I am at least on the right track in my usages of the compiler APIs.

Looks reasonable. Couple of small observations:
* You probably don't need `getSIMDVectorRegisterByteLength`. The nodes should already have the correct SIMD types (you may want to assert that it's TYP_SIMD16/32 since there's no bitwise AND for TYP_SIMD8/12) and you can use the type to determine which HW intrinsic to use. It's unfortunate that the HW intrinsic design requires this special casing, oh well...
* You should probably use `SSE41_TestZ` instead of `SSE41_TestAllZeros` for consistency with AVX. No idea why `SSE41_TestAllZeros` exists.
* The code can be moved outside of `if (TryGetUse)`, dealing with SETCC/JCC will become `LowerHWIntrinsic`'s responsibility.

> There is no HW-intrinsic we can use directly for SIMDOpInEquality (I think?)

Right, interesting situation. With 17073 it shouldn't be an issue to handle this but it probably means that we won't be able to use `LowerHWIntrinsic` or that we perhaps need to add `TestNZ` to do so.

**Comment by jakobbotsch** [comment]
> Looks reasonable. Couple of small observations:

Thanks. During this I have also noticed that the current `ptest` optimization does not apply when the operands are flipped, i.e. `Vector<byte>.Zero == (a & b)` generates
```asm
; Assembly listing for method Program:TestMask(byref,byref):bool
; Emitting BLENDED_CODE for X64 CPU with AVX
; optimized code
; rsp based frame
; partially interruptible
; Final local variable assignments
;
;  V00 arg0         [V00,T00] (  3,  3   )   byref  ->  rcx
;  V01 arg1         [V01,T01] (  3,  3   )   byref  ->  rdx
;# V02 OutArgs      [V02    ] (  1,  1   )  lclBlk ( 0) [rsp+0x00]
;
; Lcl frame size = 0

G_M55725_IG01:
       C5F877               vzeroupper

G_M55725_IG02:
       C4E17D1001           vmovupd  ymm0, ymmword ptr[rcx]
       C4E17D100A           vmovupd  ymm1, ymmword ptr[rdx]
       C4E17DDBC1           vpand    ymm0, ymm1
       C4E175EFC9           vpxor    ymm1, ymm1
       C4E17C28D1           vmovaps  ymm2, ymm1
       C4E16D76D0           vpcmpeqd ymm2, ymm0
       C4E17DD7C2           vpmovmskb eax, ymm2
       83F8FF               cmp      eax, -1
       0F94C0               sete     al
       0FB6C0               movzx    rax, al
       0FB6C0               movzx    rax, al

G_M55725_IG03:
       C5F877               vzeroupper
       C3                   ret

; Total bytes of code 54, prolog size 3 for method Program:TestMask(byref,byref):bool
; ============================================================
```
Probably not super important as most people use the first variant, but I imagine it could turn up due to inlining or other optimizations.

> With 17073 it shouldn't be an issue to handle this but it probably means that we won't be able to use LowerHWIntrinsic or that we perhaps need to add TestNZ to do so.

When is it appropriate to add node flags? I have noticed that there is a `GTF_JCMP_EQ` flag to handle a similar situation to this in ARM code-gen.

**Comment by mikedn** [comment]
> When is it appropriate to add node flags? I have noticed that there is a GTF_JCMP_EQ flag to handle a similar situation to this in ARM code-gen.

Personally I prefer to avoid node flags, they tend to increase code complexity and the chance of introducing bugs (there are already many flags and it's not that difficult to forget to check one). But sometimes they're necessary, at least due to shortcomings of the JIT's IR.

But perhaps we won't need neither flags nor new intrinsics. The experimental implementation I've done for 17073 adds a `LowerHWIntrinsicCC` that takes a condition, perhaps we could simply call that with the appropriate condition code. I don't know right now, it's been a while since I worked on that.

https://github.com/mikedn/coreclr/blob/c4e25a760b93113489d3689d22a0815e26b23b5a/src/jit/lowerxarch.cpp#L811

**Comment by erozenfeld** [comment]
@dotnet/jit-contrib

**Comment by CarolEidt** [comment]
> The experimental implementation I've done for 17073 adds a LowerHWIntrinsicCC that takes a condition, perhaps we could simply call that with the appropriate condition code.

That seems like the right direction - is this something you plan to pursue?

@jakobbotsch - it looks like you've got a good start.

**Comment by mikedn** [comment]
> The key issues being that SIMDOpInEquality is not supported, and the duplicate movzx

That duplicate `movzx` that appears in some case is inserted by morph when normalizing small type values is deemed necessary (e.g. when the return type of the method is `bool`). `fgCastNeeded` has special handling that avoids inserting the cast for relops but it does not know about SIMD intrinsics.

Interestingly, this does not happen for HW intrinsics. `bool` returning intrinsics are (wrongly?) imported using `TYP_BOOL` so a cast is not inserted.
